### PR TITLE
Implement CRUD tests for API Hosts

### PR DESCRIPTION
**Pull request by abalakh**
```python
py.test -v tests/foreman/api/test_host.py
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1
collected 17 items 

tests/foreman/api/test_host.py::HostTestCase::test_negative_update_arch <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_mac <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_name <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_os <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_name <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_owner_type <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_delete <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_get_per_page <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_get_search <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_arch <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_domain <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_env <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_mac <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_medium <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_name <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_os <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_owner_type <- robottelo/decorators.py PASSED

============================ 17 passed in 366.23 seconds =============================
```

### CODE REVIEW

File: tests/foreman/api/test_host.py [hunk 1]
```diff
@@ -5,10 +5,12 @@
 http://theforeman.org/api/apidoc/v2/hosts.html
 
 """
-from fauxfactory import gen_integer, gen_string
+from fauxfactory import gen_integer, gen_mac, gen_string
 from nailgun import client, entities
+from requests.exceptions import HTTPError
 from robottelo.config import settings
-from robottelo.decorators import bz_bug_is_open, run_only_on, tier1
+from robottelo.datafactory import valid_hosts_list, invalid_values_list
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: `invalid_values_list, valid_hosts_list`

File: tests/foreman/api/test_host.py [hunk 2]
```diff
@@ -91,3 +93,244 @@ def test_positive_update_owner_type(self):
                 host.owner_type = owner_type
                 host = host.update(['owner_type'])
                 self.assertEqual(host.owner_type, owner_type)
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_name(self):
+        """Create a host with a random valid name
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: It seems that we should have plural form here

File: tests/foreman/api/test_host.py [hunk 3]
```diff
@@ -91,3 +93,244 @@ def test_positive_update_owner_type(self):
...
+
+        @feature: Hosts
+
+        @assert: A host is created and the name matches
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: Not so obvious description. I mean that `matches` word

File: tests/foreman/api/test_host.py [hunk 4]
```diff
@@ -91,3 +93,244 @@ def test_positive_update_owner_type(self):
...
+        """
+        for name in valid_hosts_list():
+            with self.subTest(name):
+                host = entities.Host()
+                host.create_missing()
+                host.name = name
+                host = host.create(create_missing=False)
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: What is a specific reason to not have `host = entities.Host(name=name).create`?
Anyway that way of implementation doesn't seem similar to the code that we are usually use. Could you fix that?

**Comment by oshtaier** [comment]
Can we add few more cases for `create` functionality with additional entities like `domain` or `architecture`?

**Comment by oshtaier** [comment]
Pending comments

File: tests/foreman/api/test_host.py (hunk 1 above)
Comment (abalakh) [inline tests/foreman/api/test_host.py]: :+1:

File: tests/foreman/api/test_host.py (hunk 4 above)
Comment (abalakh) [inline tests/foreman/api/test_host.py]: That was done on purpose. Take a look at [Host.create_missing()](https://github.com/SatelliteQE/nailgun/blob/master/nailgun/entities.py#L2031-L2093). Rules for creating missing entities for hosts are too complex and `create_missing()` works only for new instances of host, which do not have any instance attributes set. More information is in [Host.create_missing()](https://github.com/SatelliteQE/nailgun/blob/master/nailgun/entities.py#L2032-L2056) docstring. That's why we're creating dependencies at first, then updating required fields for future host and only then actually creating a host. Btw, same workaround is also used in CLI hosts tests.

**Comment by abalakh** [comment]
@oshtaier sure, will add some more `create` tests

**Comment by abalakh** [comment]
@oshtaier actually create tests already have all the minimum-required entities included. I mean each time we create a host it will create `architecture`, `medium`, `domain`, `os` etc and use them while creating a new host, so there's no point in separate `test_positive_create_with_domain`-like tests.

**Comment by oshtaier** [comment]
@abalakh As far as I see, create_missing will populate only few entities for `Hosts`. Right? Only required ones?

**Comment by abalakh** [comment]
@oshtaier right. they are: architecture, medium, domain, os, ptable, environment, location, organization. That's a half of them, remaining ones are: parent, puppet_ca_proxy, subnet, realm, puppet_proxy, content_source, content_view, lifecycle_environment.
Do you want me to cover them in this PR or maybe in a separate one?

**Comment by oshtaier** [comment]
@abalakh I don't think that we are in hurry with these tests, so it will be fine to have as more coverage as we can in that PR

**Comment by oshtaier** [comment]
which are not so hard to add probably :)

**Comment by omaciel** [comment]
@abalakh fwiw I played with creating `entities.Host` entities yesterday and I was able to create them slightly different:

```python
from fauxfactory import gen_mac

# Extract the server domain from the configuration object
DOMAIN = server_config.url[server_config.url.find('.')+1:]

# Fetch the default organization
results = entities.Organization().search(
    query={'search': 'name="Default Organization"'},)
assert len(results) == 1
org = results[0]

# Fetch the default content view
results = entities.ContentView(
    organization=org,
    name='Default Organization View'
).search()
assert len(results) == 1
cv = results[0]

# Fetch the 'Library' lifecycle environment
results = entities.LifecycleEnvironment(
    organization=org
).search(query={'search': 'name="Library"'},)
assert len(results) == 1
lce = results[0]

# Fetch the default location
results = entities.Location(
    organization=[org]).search(query={'search': 'name="Default Location"'})
assert len(results) == 1
loc = results[0]

# Fetch a valid architecture
results = entities.Architecture().search(query={'search': 'name="x86_64"'})
assert len(results) == 1
arch = results[0]

# Fetch the domain
results = entities.Domain().search(
    query={'search': 'name="{0}"'.format(DOMAIN)})
assert len(results) == 1
domain = results[0]

# Fetch an operating system
results = entities.OperatingSystem().search(
    query={'search': 'title="RedHat 7.2"'})
assert len(results) == 1
opsys = results[0]

# Partition table
results = entities.PartitionTable().search(
    query={'search': 'family="{0}"'.format(opsys.family)})
assert len(results) == 1
ptable = results[0]

# Media
results = entities.Media().search(query={'search': 'path~"7_2"'})
assert len(results) == 1
media = results[0]

# Environment
results = entities.Environment().search(query={'search': 'name="production"'})
assert len(results) == 1
environment = results[0]

host = entities.Host(
    environment=environment,
    organization=org,
    location=loc,
    domain=domain,
    name=u'fakesystem001',
    operatingsystem=opsys,
    architecture=arch,
    ptable=ptable,
    mac=gen_mac(),
    medium=media,
    root_pass=u'asdfasdf')
host.create(create_missing=False)
```

**Comment by abalakh** [comment]
@omaciel thanks for sharing!
Btw, why do you search for CV and LCE? They are not accepted for host creation and i can't see you used them anywhere else. Without CV and LCE - that's basically the same as `create_missing()` is providing, except the fact that `create_missing()` creates new entities and your approach just uses existing ones.

**Comment by omaciel** [comment]
@abalakh ahh those were left over from another section of code, which I did not share here, which created content hosts via API.

File: tests/foreman/api/test_host.py (hunk 3 above)
Comment (elyezer) [inline tests/foreman/api/test_host.py]: A host is created with the expected name?

File: tests/foreman/api/test_host.py [hunk 5]
```diff
@@ -91,3 +93,244 @@ def test_positive_update_owner_type(self):
...
+        """
+        host = entities.Host().create(create_missing=True)
+        host.delete()
+        with self.assertRaises(HTTPError):
+            host.read()
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_update_name(self):
+        """Update a host with a new random name
```
Comment (sthirugn) [inline tests/foreman/api/test_host.py]: please remove `random` here, we always use random test data.

**Comment by sthirugn** [comment]
ACK pending @oshtaier comments.  These are some nice tests btw.

**Comment by abalakh** [comment]
Comments addressed. Added a lot of new CRUDs, please take a look at PR once again.
Tests results:
```python
py.test -v tests/foreman/api/test_host.py
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1
collected 43 items 

tests/foreman/api/test_host.py::HostTestCase::test_negative_update_arch <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_mac <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_name <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_os <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_build_parameter <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_comment <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_compresource <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_compute_profile <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_enabled_parameter <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_hostgroup <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_ip <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_managed_parameter <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_model <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_name <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_owner_type <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_ca_proxy <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_proxy <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_subnet <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_user <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_delete <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_get_per_page <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_get_search <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_arch <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_build_parameter <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_comment <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_compresource <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_compute_profile <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_domain <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_enabled_parameter <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_env <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_hostgroup <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_ip <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_mac <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_managed_parameter <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_medium <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_model <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_name <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_os <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_owner_type <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_ca_proxy <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_proxy <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_subnet <- robottelo/decorators.py PASSED
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_user <- robottelo/decorators.py PASSED

============================ 43 passed in 1523.90 seconds ============================
```

**Comment by abalakh** [comment]
Following fields were still not covered: `compute_attributes`, `capabilities`, `progress_report_id`, `provision_method`, `image_id`, `host_parameters_attributes`, `realm_id`, `puppet_class_ids`, `owner_type`.
Some of them are pretty unfamiliar for me and/or require some infrastructure changes (e.g. `owner_type` requires changes in nailgun, right now it's not possible to cover it), so i'll try to address them in a separate PR.

**Comment by oshtaier** [comment]
Had discussions with @abalakh and agreed to separate the efforts in different PR

File: tests/foreman/api/test_host.py [hunk 6]
```diff
@@ -91,3 +97,707 @@ def test_positive_update_owner_type(self):
...
+    def test_positive_create_with_compute_profile(self):
+        """Create a host with a compute profile specified
+
+        @feature: Hosts
+
+        @assert: A host is created with expected compute profile assigned
+        """
+        host = entities.Host()
+        host.create_missing()
+        cprofile = entities.ComputeProfile().create()
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: `profile = `

File: tests/foreman/api/test_host.py [hunk 7]
```diff
@@ -91,3 +97,707 @@ def test_positive_update_owner_type(self):
...
+        user = entities.User().create()
+        host.owner_type = 'User'
+        host.owner = user
+        host = host.create(create_missing=False)
+        self.assertEqual(host.owner.read().login, user.login)
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_build_parameter(self):
+        """Create a host with 'build' parameter specified
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: Can we have a very brief explanation what 'build' parameter is?

File: tests/foreman/api/test_host.py [hunk 8]
```diff
@@ -91,3 +97,707 @@ def test_positive_update_owner_type(self):
...
+        host = entities.Host()
+        host.create_missing()
+        host.build = True
+        host = host.create(create_missing=False)
+        self.assertEqual(host.build, True)
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_enabled_parameter(self):
+        """Create a host with 'enabled' parameter specified
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: Can we have a very brief explanation what 'enabled' parameter is?

File: tests/foreman/api/test_host.py [hunk 9]
```diff
@@ -91,3 +97,707 @@ def test_positive_update_owner_type(self):
...
+        host = entities.Host()
+        host.create_missing()
+        host.enabled = False
+        host = host.create(create_missing=False)
+        self.assertEqual(host.enabled, False)
+
+    @run_only_on('sat')
+    @tier1
+    def test_positive_create_with_managed_parameter(self):
+        """Create a host with managed parameter specified
```
Comment (oshtaier) [inline tests/foreman/api/test_host.py]: Can we have a very brief explanation what 'managed' parameter is?

**Comment by oshtaier** [comment]
ACK pending non-blocking comments
and please provide new issue id for things that we discussed here (like what parameters are left to cover and 'usergroup specifics')

**Comment by abalakh** [comment]
@oshtaier comments addressed. Issue for missing fields is #3270.
For usergroups we already have SatelliteQE/nailgun#257 PR, and open issue SatelliteQE/nailgun#258

**Comment by oshtaier** [comment]
ACK

**Pull request merged**